kernel: Use spans instead of vectors for passing block headers to validation functions

sedited opened the pull request:
Makes it friendlier for potential future users of the kernel library if they do not store the headers in a std::vector, but can guarantee contiguous memory.

Take this opportunity to also change the argument of ImportBlocks previously taking a `std::vector` to a `std::span`.

sedited committed:
validation: Use span for ProcessNewBlockHeaders

Makes it friendlier for potential future users of the kernel library if
they do not store the headers in a std::vector, but can guarantee
contiguous memory.

sedited committed:
validation: Use span for CalculateClaimedHeadersWork

Makes it friendlier for potential future users of the kernel library if
they do not store the headers in a std::vector, but can guarantee
contiguous memory.

theuni commented on src/validation.cpp:
```diff
@@ -4384,7 +4385,7 @@ bool ChainstateManager::AcceptBlockHeader(const CBlockHeader& block, BlockValida
 
 // Exposed wrapper for AcceptBlockHeader
-bool ChainstateManager::ProcessNewBlockHeaders(const std::vector<CBlockHeader>& headers, bool min_pow_checked, BlockValidationState& state, const CBlockIndex** ppindex)
+bool ChainstateManager::ProcessNewBlockHeaders(const std::span<CBlockHeader> headers, bool min_pow_checked, BlockValidationState& state, const CBlockIndex** ppindex)
```
I think you want `std::span<const CBlockHeader>` for these. In addition to preserving constness, it allows elements to be passed in as `{{foo}}`, rather than specifying spans everywhere.

theuni reviewed (commented):
Concept ACK.

theuni commented:
Actually, I read this assuming we'd already migrated away from `Span`, which isn't the case. So probably best not to start introducing `std::span`s until that's happened?

theuni commented:
Relevant: #29119

TheCharlatan commented:
Thanks for the review @theuni!

Updated 65e440433e7223511a4effb568cdb187fa0aa22c -> c584c1f23b8107dade11d81ada2baf741acbac62 ([headersSpan_0](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_0) -> [headersSpan_1](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/headersSpan_0..headersSpan_1))

* Addressed @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/30742#discussion_r1735049626), made the span take a `const CBlockHeader`, which allows dropping the annoying `std::span` invocations again.

> Actually, I read this assuming we'd already migrated away from Span, which isn't the case. So probably best not to start introducing std::spans until that's happened?

There already are users of `std::span`, so I did not have the impression that I was introducing something novel here. Now that #29071 is merged I would also expect #29119 to move forward soon. But can always revert to `Span` if that is what reviewers prefer.

theuni commented on src/validation.cpp:
```diff
@@ -4384,7 +4385,7 @@ bool ChainstateManager::AcceptBlockHeader(const CBlockHeader& block, BlockValida
 
 // Exposed wrapper for AcceptBlockHeader
-bool ChainstateManager::ProcessNewBlockHeaders(const std::vector<CBlockHeader>& headers, bool min_pow_checked, BlockValidationState& state, const CBlockIndex** ppindex)
+bool ChainstateManager::ProcessNewBlockHeaders(const std::span<const CBlockHeader> headers, bool min_pow_checked, BlockValidationState& state, const CBlockIndex** ppindex)
```
No need to pass as const. 
`foo(std::span<const CBlockHeader>)` is functionally equivalent to `foo(const std::vector<CBlockHeader>&)`.

TheCharlatan replied:
Whoops, missed that, thanks!

theuni commented:
> Now that #29071 is merged I would also expect #29119 to move forward soon. But can always revert to `Span` if that is what reviewers prefer.

@maflcko  ?

TheCharlatan commented:
Updated c584c1f23b8107dade11d81ada2baf741acbac62 -> 45a90050290b046a8c0f417eab679f36032da297 ([headersSpan_1](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_1) -> [headersSpan_2](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/headersSpan_1..headersSpan_2))

* Addressed @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/30742#discussion_r1735095909), dropped superfluous const.

jonatack reviewed (commented):
ACK 45a90050290b046a8c0f417eab679f36032da297

TheCharlatan commented:
Thanks for the review @jonatack.

Updated 45a90050290b046a8c0f417eab679f36032da297 -> 797eede7609ca0d4d788e62f35e70e85103a19dc ([headersSpan_2](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_2) -> [headersSpan_3](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/headersSpan_2..headersSpan_3))

* Addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/30742#discussion_r1735122911), dropped now unneeded temporaries.

stickies-v reviewed (approved):
ACK 797eede7609ca0d4d788e62f35e70e85103a19dc

TheCharlatan commented:
>  but I don't think any compiler implements it and the code here is compiled in C++20, so I haven't checked if that fixes your problem.

Thanks, you are right, my initial patch avoided copies where possible by explicitly creating the span, but the brace initializers that theuni [suggested](https://github.com/bitcoin/bitcoin/pull/30742#discussion_r1735049626) instead create temporary arrays that copy the headers. I should update the description, or I can revert to what I was doing before. Do you have a preference? I don't think the copies matter all that much either way. The main motiviation for me is that it makes life a bit easier for potential external callers if you use spans over vector references in the interfaces.

sedited committed:
validation: Use span for ImportBlocks paths

Makes it friendlier for potential future users of the kernel library if
they do not store the headers in a std::vector, but can guarantee
contiguous memory.

stickies-v commented on src/node/blockstorage.h:
```diff
@@ -429,7 +430,7 @@ class BlockManager
 };
 
-void ImportBlocks(ChainstateManager& chainman, std::vector<fs::path> vImportFiles);
+void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> vImportFiles);
```
nit: would be nice to modernize the name to e.g. `import_paths` since it's no longer a vector.
```suggestion
void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> import_paths);
```

TheCharlatan replied:
Oh right, finally!

stickies-v reviewed (approved):
ACK b31483e32b7d111eeef5fea8ebdc927a32c22bf4

TheCharlatan commented:
Updated b31483e32b7d111eeef5fea8ebdc927a32c22bf4 -> bf34edfbc38079080f007feb6bd532471707fc65 ([headersSpan_4](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_4) -> [headersSpan_5](https://github.com/TheCharlatan/bitcoin/tree/headersSpan_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/headersSpan_4..headersSpan_5))

* Modernizing naming as suggested by @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/30742#discussion_r1738374489).

stickies-v commented:
re-ACK a2955f09792b6232f3a45aa44a498b466279a8b7 - no changes except further walking the file of modernizing variable names.

danielabrozzoni reviewed (approved):
ACK a2955f09792b6232f3a45aa44a498b466279a8b7

achow101 commented:
ACK a2955f09792b6232f3a45aa44a498b466279a8b7
